### DISCUSSION TITLE
Route Formats in v5

### DISCUSSION THREAD

**Discussion by hugoaboud** [adonisjs/core#3059] [Help]
I'm building an API that must be compatible (for now) with some of our legacy API standards, such as the `.json` suffix on CRUD routes.

I've found the [legacy](https://legacy.adonisjs.com/docs/4.1/routing#_route_formats) solution, however I wasn't able to find it on v5 neither on the discussions/issues.
What is the recommended way to approach that on the new version?

**Comment by anthrogan** [comment]
```js
Route.get('/', () => {
  /** JSON response */
  return { page: 'home' }
})
```

or

```js
Route.get('/', ({ response }) => {
  response.json(someObj)
})
```
or

```js
Route.get('/', ({ response }) => {
  response.send({ foo: 'bar' })
})
```

https://docs.adonisjs.com/guides/response#sending-the-response

**Reply by hugoaboud** [reply]
Thank you for the quick reply. However your solution relates to the response type, which I understand is what the `.json` is supposed to do (and was probably removed due to this information being sent on the Content-Type header).
My problem lies on the fact that, given this route:

```typescript
Route.get('/test', () => {
  return { page: 'test' }
});
```

I'm not able to access it through `GET /test.json`, only through `GET /test`.
I should also be able to access endpoints with an id such as `GET /test/:id` through `GET /test/3.json` for example.

I know this is and old/bad standard, however for compatibility reasons I must implement it for now.

**Reply by anthrogan** [reply]
Oh I see,

If you are using something like `/test/data` which returns a web page for example aswell as `/test/data.json` which returns json I guess though could use param maching

https://docs.adonisjs.com/guides/routing#params-matchers

You could even group your routes by `.json` response for example
```js
Route.group( // your routes in here ).where('id', /W*(.json)\W*/gm)
```
So when a .json request is caught it runs specific end points

Edit: And for specific return all end points just make `GET /test/alldata.json` and it will match

**Reply by hugoaboud** [reply]
Thank you! The regex solved it very elegantly.

Here's my full solution:
- Routes with parameter: match optional '.json' with regex and remove it on cast.
- Routes without parameter: make a duplicate

```typescript
const id_cast = { match: /^\d+(.json)?$/, cast: (id) => Number(id.replace('.json','')) };

Route.group(() => {
    Route.post("/", this.name+".create");
    Route.get("/", this.name+".index");
    Route.get("/:id", this.name+".read");
    Route.put("/:id", this.name+".update");
    Route.delete("/:id", this.name+".delete");
}).prefix(endpoint).middleware('auth').where('id', id_cast);

Route.post("/"+endpoint+'.json', this.name+".create").middleware('auth');
Route.get("/"+endpoint+'.json', this.name+".index").middleware('auth');
``` 

(Apparently I'm not able to mark your latest reply as the topic answer, only the original one. What do you suggest?)

**Reply by anthrogan** [reply]
Makes **way** more sense now that I can see how you have used `id_cast` 😂 glad we've found a work around!

**Comment by RomainLanz** [comment]
Hey @hugoaboud! 👋 

You can still use the [Content Negotiation](https://docs.adonisjs.com/guides/request#content-negotiation). Simply create a middleware that set the `Accept` header depending on the URL received.

```
/example.json => "Accept: application/json"
/example => "Accept: text/html"
````

**Reply by Kuzzy** [reply]
@RomainLanz But there are no methods to set or modify request headers https://docs.adonisjs.com/guides/request

**Reply by Kuzzy** [reply]
It can be done like `request.request.headers.accept = 'text/csv'` But would it be better to add appropriate method(s) on `request` object?
